src/routes/getFiles.ts

```diff
@@ -0,0 +1,246 @@
+  let repoId: string
+  let ignoreTimestamps: boolean = false
+
+  // Validate request body
+  try {
+    body = asGetFilesBody(req.body)
+    repoId = body.repoId
+
+    if (body.ignoreTimestamps !== undefined) {
+      ignoreTimestamps = body.ignoreTimestamps
+    }
+
+    if (repoId === '') {
+      throw new Error(`Missing repoId.`)
+    }
+  } catch (err) {
+    throw makeApiClientError(400, err.message)
+  }
+
+  const responsePaths: GetFilesResponsePaths = {}
+
+  const bodyPaths = Object.keys(body.paths)
+
+  // Use max page size config to limit the paths processed
+  if (bodyPaths.length > maxPageSize) {
+    throw makeApiClientError(
+      422,
+      `Too many paths. maxPageSize is ${maxPageSize}`
+    )
+  }
+
+  const bodyKeys = bodyPaths.map(filePath => `${repoId}:${filePath}`)
+  // Maps a directory key to a file path for timestamp lookup
+  const filePathMapOfParentDirKey: { [dirKey: string]: string[] } = {}
+
+  // Fetch Paths:
+
+  const fetchPathsResults = await dataStore.fetch({ keys: bodyKeys })
+
+  for (const row of fetchPathsResults.rows) {
+    const documentKey = row.key
+    // No need to check for undefined case here because of how keys
+    // are defined for the fetch.
+    const documentPath = documentKey.split(':')[1]
+
+    const sentTimestamp = body.paths[documentPath]
+
+    if ('doc' in row) {
+      let fileDocument: StoreFileDocument | undefined
+      let directoryDocument: StoreDirectoryDocument | undefined
+
+      try {
+        // For file document
+        fileDocument = asStoreFileDocument(row.doc)
+      } catch (err) {
+        try {
+          // For directory document
+          directoryDocument = asStoreDirectoryDocument(row.doc)
+        } catch (err) {
+          throw makeApiClientError(
+            422,
+            `File '${row.key}' is not a file or directory.`
+          )
+        }
+      }
+
+      // Handle file document
+      if (fileDocument !== undefined) {
+        responsePaths[documentPath] = {
+          ...asStoreFile(fileDocument),
+          timestamp: 0 // Timestamp will be retrieve from parent directory doc
+        }
+
+        // Add file path to file directory map for timestamp lookup
+        const [parentDirectoryPath = '/'] = getParentPathsOfPath(documentPath)
+        const parentDirectoryKey = `${repoId}:${parentDirectoryPath}`
+
+        if (filePathMapOfParentDirKey[parentDirectoryKey] === undefined) {
+          filePathMapOfParentDirKey[parentDirectoryKey] = [documentPath]
+        } else {
+          filePathMapOfParentDirKey[parentDirectoryKey].push(documentPath)
+        }
+      }
+
+      // Handle directory document
+      if (directoryDocument !== undefined) {
+        const storeDirectory = asStoreDirectory(directoryDocument)
+        const allTimestamps = [
+          ...Object.values(storeDirectory.deleted),
+          ...Object.values(storeDirectory.paths)
+        ]
+        const directoryTimestamp = Math.max(...allTimestamps)
```

Comment:
Technically, the Directory timestamp is in its parent. That's the value that should be compared

Reply:
It seems like the batch fetch of documents on line 94 should include all of the requested paths as well as their immediate parent directories. No need to accumulate the parent paths and then fetch them later as you know you will need them anyway.

Reply:
> Technically, the Directory timestamp is in its parent. That's the value that should be compared

Specs mention the directory's timestamp is the max of its "pointers". If you think about it, the this max should be equal to parent directory's pointer for a particular directory because a directory is only updated when one of it's pointers is updated.

Reply:
> It seems like the batch fetch of documents on line 94 should include all of the requested paths as well as their immediate parent directories. No need to accumulate the parent paths and then fetch them later as you know you will need them anyway.

You're right. Originally, I there was a condition in place for including the parent path in the subsequent query, but now that this condition is removed, it makes sense to only have one query. I'll make a note.

Reply:
Yes but when pulling the time stamp from the parent you can do it in constant Time vs iterating over all entries in a directory

Reply:
And since you know that you will need the parent directories anyway might as well look up the time stamp in there

Reply:
Ah, but currently the parent directory for a directory document is not queried. Only parent directories for file documents are later queried. Best performance would be to include the timestamp information in the file and directory document's themselves, eliminating the need for an extra query or calculating a maximum over a set.

Reply:
As we discussed, putting the time stamp in the file or directory Instead of the parent might speed up get files but would significantly slow down get updates which is called far more often

Reply:
And in the case where we support deleting directories, you will then need to fetch the parents of directories

Reply:
> As we discussed, putting the time stamp in the file or directory Instead of the parent might speed up get files but would significantly slow down get updates which is called far more often

I was proposing to include a file/directory's timestamp in the file/directory's document in addition, not instead. Though, having this redundant data might have some risk of data integrity issues. All in all, best to not pre-optimize here, perhaps.